### ISSUE TITLE
The scopes of traverse_mut are empty.

### ISSUE THREAD

**Issue by anguish-huihuang** [oxc-project/oxc#8087]
I have defined a HandleVariableDeclaration structure and employed traverse_mut to iterate through the nodes in the program. Within the HandleVariableDeclaration structure, I used Traverse and created the enter_program function, where I defined two new structures. If I wish to continue traversing the nodes using either the traverse_mut_with_ctx method or the traverse_mut method, I need to extract scopes and symbols from ctx. This situation results in ctx.scoping() being empty within the HandleVariableDeclaration structure, and the scopes returned by the external traverse_mut are also empty. How can this issue be resolved?
![微信图片_20241224110102](https://github.com/user-attachments/assets/a0ebc9aa-093c-437c-8353-ba5cb4ffb3ea)
![微信图片_20241224110055](https://github.com/user-attachments/assets/54dfd83c-d897-4cce-911e-40b9bb4c7e8f)

**Comment by Boshen** [comment]
You need to run `oxc_semantic` prior to traverse.

**Comment by anguish-huihuang** [comment]
Yes, I had run semantic analysis before executing Traverse, and the issue arises because in the enter_program method of the HandleVariableDeclaration structure, I extracted the scopes from ctx and assigned them to a new variable. This resulted in ctx.scopes being empty. The current solution is to convert ReusableTraverseCtx back into TraverseCtx and reassign it to ctx at the end. Is there perhaps a better method?
![微信图片_20241224123454](https://github.com/user-attachments/assets/2d1112ac-5a1a-4678-b078-66701bba27a8)
![微信图片_20241224123459](https://github.com/user-attachments/assets/20c8d314-8e21-43b0-8e44-64a8045d8e23)

**Comment by overlookmotel** [comment]
I suggest you look at the transformers we've built in `oxc_transformer` crate and look at how we compose different transforms together into a single AST pass. The code you've posted is quite inefficient, as well as wildly unsafe! `ReusableTraverseCtx::unwrap` is an unsafe method for a reason! Please don't use it unless you want to lose all your hair in a freak tornado.
